### PR TITLE
feat: use gRPC Bidirectional Streaming for Mapper and Transformer

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: src/main/java/io/numaproj/numaflow/mapper/MapSupervisorActor.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import akka.japi.pf.ReceiveBuilder;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import io.numaproj.numaflow.map.v1.MapOuterClass;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/mapper/MapSupervisorActor.java:19]: 👍

File: src/main/java/io/numaproj/numaflow/mapper/Service.java [hunk 2]
```diff
@@ -1,59 +1,90 @@
...
-        MessageList messageList = mapper.processMessage(request
-                .getKeysList()
-                .toArray(new String[0]), handlerDatum);
-
-        // set response
-        responseObserver.onNext(buildResponse(messageList));
-        responseObserver.onCompleted();
+        CompletableFuture<Void> failureFuture = new CompletableFuture<>();
+
+        handleFailure(failureFuture);
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/mapper/Service.java:48]: why do we handle a failure future right after creating it?

File: src/test/java/io/numaproj/numaflow/mapper/ServerTest.java [hunk 3]
```diff
@@ -57,31 +59,86 @@ public void tearDown() throws Exception {
...
+
+        var stub = MapGrpc.newStub(inProcessChannel);
+        var requestStreamObserver = stub
+                .mapFn(responseObserver);
+
+        requestStreamObserver.onNext(inDatum);
+
+        try {
+            responseObserver.done.get();
+        } catch (InterruptedException | ExecutionException e) {
```
Comment (KeranYang) [inline src/test/java/io/numaproj/numaflow/mapper/ServerTest.java:136]: nit: if there was no exception thrown, this unit test could pass. consider returning within the catch block and failing if reaching the end of the test.

File: src/main/java/io/numaproj/numaflow/mapper/MapSupervisorActor.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        log.debug("got a dead letter, stopping the execution");
+        responseObserver.onError(Status.UNKNOWN.withDescription("dead letters").asException());
+        failureFuture.completeExceptionally(new Throwable("dead letters"));
+        getContext().getSystem().stop(getSelf());
+    }
+
+    @Override
+    public SupervisorStrategy supervisorStrategy() {
+        // we want to stop all child actors in case of any exception
+        return new AllForOneStrategy(
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/mapper/MapSupervisorActor.java:129]: this is good. is it something we can also leverage for reduce stream/session?

File: src/main/java/io/numaproj/numaflow/mapper/Service.java (hunk 2 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/mapper/Service.java:48]: we handle failure using a separate thread.

**Comment by KeranYang** [comment]
LGTM!

**Review by KeranYang — APPROVED** [review]

**Pull request merged**